libethash/internal.c

```diff
@@ -37,264 +41,456 @@
 #include "sha3.h"
 #endif // WITH_CRYPTOPP
 
-uint64_t ethash_get_datasize(const uint32_t block_number) {
-    assert(block_number / ETHASH_EPOCH_LENGTH < 2048);
-    return dag_sizes[block_number / ETHASH_EPOCH_LENGTH];
+uint64_t ethash_get_datasize(uint64_t const block_number)
+{
+	assert(block_number / ETHASH_EPOCH_LENGTH < 2048);
+	return dag_sizes[block_number / ETHASH_EPOCH_LENGTH];
 }
 
-uint64_t ethash_get_cachesize(const uint32_t block_number) {
-    assert(block_number / ETHASH_EPOCH_LENGTH < 2048);
-    return cache_sizes[block_number / ETHASH_EPOCH_LENGTH];
+uint64_t ethash_get_cachesize(uint64_t const block_number)
+{
+	assert(block_number / ETHASH_EPOCH_LENGTH < 2048);
+	return cache_sizes[block_number / ETHASH_EPOCH_LENGTH];
 }
 
 // Follows Sergio's "STRICT MEMORY HARD HASHING FUNCTIONS" (2014)
 // https://bitslog.files.wordpress.com/2013/12/memohash-v0-3.pdf
 // SeqMemoHash(s, R, N)
-void static ethash_compute_cache_nodes(
-        node *const nodes,
-        ethash_params const *params,
-        const uint8_t seed[32]) {
-    assert((params->cache_size % sizeof(node)) == 0);
-    uint32_t const num_nodes = (uint32_t) (params->cache_size / sizeof(node));
-
-    SHA3_512(nodes[0].bytes, seed, 32);
-
-    for (unsigned i = 1; i != num_nodes; ++i) {
-        SHA3_512(nodes[i].bytes, nodes[i - 1].bytes, 64);
-    }
-
-    for (unsigned j = 0; j != ETHASH_CACHE_ROUNDS; j++) {
-        for (unsigned i = 0; i != num_nodes; i++) {
-            uint32_t const idx = nodes[i].words[0] % num_nodes;
-            node data;
-            data = nodes[(num_nodes - 1 + i) % num_nodes];
-            for (unsigned w = 0; w != NODE_WORDS; ++w) {
-                data.words[w] ^= nodes[idx].words[w];
-            }
-            SHA3_512(nodes[i].bytes, data.bytes, sizeof(data));
-        }
-    }
-
-    // now perform endian conversion
-#if BYTE_ORDER != LITTLE_ENDIAN
-    for (unsigned w = 0; w != (num_nodes*NODE_WORDS); ++w)
-    {
-        nodes->words[w] = fix_endian32(nodes->words[w]);
-    }
-#endif
-}
-
-void ethash_mkcache(
-		void *cache,
-        ethash_params const *params,
-        const uint8_t seed[32]) {
-	node *nodes = (node *) cache;
-    ethash_compute_cache_nodes(nodes, params, seed);
+bool static ethash_compute_cache_nodes(
+	node* const nodes,
+	uint64_t cache_size,
+	ethash_h256_t const* seed
+)
+{
+	if (cache_size % sizeof(node) != 0) {
+		return false;
+	}
+	uint32_t const num_nodes = (uint32_t) (cache_size / sizeof(node));
+
+	SHA3_512(nodes[0].bytes, (uint8_t*)seed, 32);
+
+	for (uint32_t i = 1; i != num_nodes; ++i) {
+		SHA3_512(nodes[i].bytes, nodes[i - 1].bytes, 64);
+	}
+
+	for (uint32_t j = 0; j != ETHASH_CACHE_ROUNDS; j++) {
```

Comment:
braces on next line, a lot of braces style errors in this file, also check rest of file.